lib/build.ts

```diff
@@ -187,3 +200,125 @@ export async function runBuildTask(
+//   In such csaes, Docker will try to guess based on the platform of the image based
+//   on the current machine arch.
+//
+//   When Docker recieves a target platform, these are the possible cases:
+//
+//      + V2 manifest for all images, and all images support the requested arch
+//        Note that "support" means either it is a single arch image with the correct arch
+//        or else it is a multi-arch image with one of the images matching arch
+//          -> Result: Good to go
+//
+//      + V2 manifest for all images, and some do not suppor the requested arch
+//          -> Result: Docker error
+//
+//      + Some V1 manifests, the plaform flag matches what Docker guesses
+//          -> Result:  Docker will succeed if the guesses are correct or fail with
+//                      execution error if one of the guesses is wrong.
+//
+//      + Some V1 manifests, platform flag does not match what Docker guesses
+//          -> Result:  Docker will throw an error even if the images are actually the
+//                      correct architecture, because it's guesses are wrong.
+//
+//   When Docker does not recieve the `plaform` flag, it sort of closes it's eyes and hopes
+//   for the bests.  It has to do this in order to maintain backwards-compatible behaviour:
+//
+//      + V2 manifests for all images. If using multiarch images, Docker will
+//        assumee the current platform.  If using single arch images, Docker will
+//        assume the declared image arch is the correct platform.
+//          -> Result: Docker builds, with a warning if the images do not match the
+//                     current platform.  Everything works as long as bin_fmt
+//                     is enabled correctly.  Otherwise, an execution error will occur.
+//
+//      + Some V1 manifests
+//          -> Result: Docker builds, assuming that the platforms will match up.
+//                     Everything works as long as bin_fmt is enabled correctly.
+//                     Otherwise, an execution error will occur.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manfifest, we can simply pass the plaform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close it's eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
+//
+async function getRecommendedPlatformhHandling(
+	task: BuildTask,
+	docker: Dockerode,
+): Promise<'allowEnginePlatformHandling' | 'noPlatformHandling' | 'unknown'> {
+	let imageReferences: string[];
+
+	const getHasV1Manfiest = async (repository: string) => {
```

Comment:
`getHas...` is a bit awkward, Java/C# is strong in you padawan :) There's also a typo.

```suggestion
	const hasV1Manifest = async (repository: string) => {
```